[Woo POS] [Cash & Receipts] Validate cash amount

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #14749 

## Description
This PR adds validation to the cash amount entered by a merchant during the cash collection flow.
- If the amount is not enough, does not show any message until we press the "collect" button, at which point an error message (red) will appear
- If the amount is exact, we can collect normally
- If the amount exceeds the order total, a "change due" (green) message will be shown each time we change the input.


https://github.com/user-attachments/assets/348d2b57-bd86-4589-8991-9b5e20c14633


## Testing
- Enable `.acceptCashForPointOfSale` feature flag
- In POS, add items to the cart, then tap on "Cash Payment"
- Enter an amount less, equal, and exceeding the order total. Observe that responds as in the description
- Using `.` or `,` for decimals should behave the same.
- Upon completion, the order appears in the app under the Orders tab.

You will see some additional space at the bottom of the success view, as well as the floating button to connect the reader. This is unrelated to this PR and will be addressed in the parent PR (#14602) or in a follow-up

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
compare order total vs merchant input to validate

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
separate change due and error strings when validating

iamgabrielma committed:
use textSuccess color for change due

iamgabrielma committed:
localized strings

iamgabrielma commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -112,6 +117,46 @@ struct PointOfSaleCollectCashView: View {
 }
 
+private extension PointOfSaleCollectCashView {
+    func parseCurrency(_ amountString: String) -> NSDecimalNumber? {
```
I've attempted to implement this from zero using `Decimal` but dealing with currency was anything but fun, so I've relied on our own `CurrencyFormatter` implementation. That said we seem to carry our own downsides (eg: https://github.com/woocommerce/woocommerce-ios/issues/13829 ) so I'm not sure if there's a better way we could use to handle formatting here, since we use the same currency formatter in other places in POS 🤔

staskus replied:
Yes, understandable. It's okay to use what we currently have. We can make improvements internally within the formatters.

The largest issues I've seen were when converting amounts back and forth when creating and then editing Order fields again. It's not the case with this POS cash payment so it should be fine.

staskus commented on WooCommerce/Classes/POS/Presentation/PointOfSaleCollectCashView.swift:
```diff
@@ -112,6 +117,46 @@ struct PointOfSaleCollectCashView: View {
+        currencyFormatter.convertToDecimal(amountString, locale: .current)
+    }
+
+    private func formatAsCurrency(_ amount: NSDecimalNumber) -> String {
```
All these methods could be unit tested (with mocked currency formatter). We have calculations happening within them. These methods would likely need to change into pure functions within a helper and then View would just set appropriate labels based on the result.

staskus reviewed (commented):
Works well, thanks!

I suggest moving cash calculations into a separate helper and to write unit tests.

iamgabrielma committed:
Wrap calculation methods in CollectCashViewHelper

iamgabrielma committed:
Extract CollectCashViewHelper to own file

iamgabrielma committed:
Add tests for updatechangeDueMessage

iamgabrielma committed:
tests for validateAmountOnSubmit

iamgabrielma commented:
Thanks for the review @staskus !

> I suggest moving cash calculations into a separate helper and to write unit tests.

I've moved the functions to a `CollectCashView` helper on https://github.com/woocommerce/woocommerce-ios/pull/14811/commits/434ebbc12166de11d6ebfca63c4bc20b8872bf07 and https://github.com/woocommerce/woocommerce-ios/pull/14811/commits/c4b01fa582228002a61fb6e8d5880869a75febb1, leaving only a small wrapper in our View on a private extension so there's no logic in the way. 

I've also added Unit Tests for those on https://github.com/woocommerce/woocommerce-ios/pull/14811/commits/46522c1a2a03adcbab75cb8d025f7fdebe421bd1 and https://github.com/woocommerce/woocommerce-ios/pull/14811/commits/61efa28e7c15716d5cd3d790a0b36b55ed97943d

For the moment I haven't injected the CurrencyFormatter and I rely on the service locator instance. Happy to iterate in this or a different PR if you think is worth it at the moment.

staskus reviewed (approved):
Thank you for the improvements! Looks good.
